### ISSUE TITLE
Round-tripping numbers from YAML to JSON does not work

### ISSUE THREAD

**Issue by nichtich** [LibreCat/Catmandu#236]
Because of inconsistencies between YAML::XS and JSON::XS, converting numbers from YAML to JSON does not work:

```
$ echo '{"n":42}' | catmandu convert JSON to YAML
---
n: 42
...
$ echo '{"n":42}' | catmandu convert JSON to YAML | catmandu convert YAML to JSON
[{"n":"42"}]
```

# Debugging

The problem is caused by JSON::XS (I tried the newest 3.02, Catmandu only requires 2.3):

```
use v5.14.1;
use YAML::XS ();
use Data::Dumper ();
use JSON::XS ();
use Devel::Peek ();

my ($yaml) = YAML::XS::Load("---\nn: 42");

say YAML::XS::Dump($yaml);
say Data::Dumper::Dumper($yaml->{n});
say Devel::Peek::Dump($yaml->{n});
say JSON::XS::encode_json($yaml); # emits string instead of integer 
say;
$yaml->{n} = 42;
say Devel::Peek::Dump($yaml->{n});
say JSON::XS::encode_json($yaml); # emits integer
```
---
```
---
n: 42

$VAR1 = 42;

SV = PVIV(0x1ba7370) at 0x1ba2788
  REFCNT = 1
  FLAGS = (IOK,POK,pIOK,pPOK)
  IV = 42
  PV = 0x1bbaa60 "42"\0
  CUR = 2
  LEN = 16

{"n":"42"}

SV = PVIV(0x1ba7370) at 0x1ba2788
  REFCNT = 1
  FLAGS = (IOK,pIOK)
  IV = 42
  PV = 0x1bbaa60 "42"\0
  CUR = 2
  LEN = 16

{"n":42}
```

Looks like the `POK` flag make JSON::XS treat the variable as string while Data::Dumper and YAML::XS::Dump treat it as number because they look at `IOK` flag and `IV` field first. Shorter analysis:

```
use v5.14.1;
use Scalar::Util qw(dualvar);
use Devel::Peek ();
use JSON::XS ();
use YAML::XS ();
use Data::Dumper ();

my $value = dualvar 42, '42';
say Devel::Peek::Dump($value);
say YAML::XS::Dump({ value => $value });        # number
say Data::Dumper::Dumper({ value => $value });  # number
say JSON::XS::encode_json({ value => $value }); # string
```

# Summary

In other words: YAML::XS reads numbers as "dual values" and writes dual values as numbers. JSON::XS writes dual values as strings (but [Mojo::JSON does not](https://github.com/kraih/mojo/issues/591)). We could 

* switch to Mojo::JSON (just joking) 
* modify Catmandu::Impoprter::YAML to check with Scalar::Util::isdual and then modify the internal representation of numbers
* modify Catmandu::Exporter::JSON to check with Scalar::Util::isdual and then modify the internal representation of numbers

What do you think?

**Comment by kraih** [comment]
Use `Cpanel::JSON::XS` instead of `JSON::XS`.

**Comment by nichtich** [comment]
Glad to see it's already in Debian (<https://packages.debian.org/jessie/libcpanel-json-xs-perl>). The version available in Ubuntu 14.04 LTS (<http://packages.ubuntu.com/trusty/libcpanel-json-xs-perl>) seems to be too old for this issue. I prefer to only require very stable XS modules because sometimes it is not possible to install them but I'm ok with migration to [Cpanel::JSON::XS](https://metacpan.org/pod/Cpanel::JSON::XS).

**Comment by nics** [comment]
switched to Cpanel::JSON::XS in 90286514cb5ef188106a0b06beab1054f7ec7a49

**Comment by nics** [comment]
roundtripping works now
